src/Microsoft.AspNetCore.Routing/EndpointDataSourceBuilder.cs

```diff
@@ -1,15 +1,18 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
 using Microsoft.AspNetCore.Builder;
 
 namespace Microsoft.AspNetCore.Routing
 {
-    public abstract class EndpointDataSourceBuilder
+    public abstract class EndpointDataSourcesBuilder
     {
-        public abstract ICollection<EndpointBuilder> Endpoints { get; }
-
         public abstract IApplicationBuilder CreateApplicationBuilder();
+
+        public abstract IServiceProvider ServiceProvider { get; }
+
+        public abstract IList<EndpointDataSource> EndpointDataSources { get; }
```

Comment:
What should change on `EndpointDataSourcesBuilder` to become `IEndpointRoutesBuilder`?

With calls like `Map`, `MapGet`, etc, the user is creating a single endpoint and there is no consideration for that endpoint ever changing.

Meanwhile a call like `MapMvc("{controller}/{action}")` will configure a datasource that has a change token.

My idea: Perhaps these methods are populated onto a single static datasource.

```cs
public interface IEndpointRoutesBuilder
{
    IApplicationBuilder CreateApplicationBuilder();
    IServiceProvider ServiceProvider { get; }
    IList<EndpointDataSource> EndpointDataSources { get; }
    IList<Endpoint> StaticEndpoints { get; }
}
```

Reply:
What is CreateApplicationBuilder for?

Reply:
Perhaps I'm missing something. What would go wrong if `Map` used the same technique as `MapMvc...`? Create and store a single data source, and then add stuff to it repeatedly?

Reply:
@davidfowl So you can easily turn existing middleware into an endpoint

```cs
public static class EndpointDataSourceBuilderExtensions
{
    public static IApplyEndpointBuilder MapHello(this EndpointDataSourcesBuilder builder, string template, string greeter)
    {
        if (builder == null)
        {
            throw new ArgumentNullException(nameof(builder));
        }

        var pipeline = builder.CreateApplicationBuilder()
           .UseHello(greeter)
           .Build();

        return builder.Map(
            template,
            "Hello " + greeter,
            pipeline);
    }
}
```

Good? Bad? Better way of doing it?

Reply:
@rynowak There is nothing wrong with doing that. To me it feels weird that this builder has a list of datasources on it, yet we aren't using it. We're just using ServiceProvider and building datasources that have been registered in DI.

Maybe MvcEndpointDataSource shouldn't be registered in DI but created and stored on this collection (i.e. get it from the list, if it is null then create the singleton instance).
Use the same approach for "static" endpoints.